File: 2024/3/solution.go
// https://adventofcode.com/2024/day/3
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"time"
)


var isTest *bool = flag.Bool("test", false, "Read the test input")


func timeTrack(start time.Time, name string) {
    elapsed := time.Since(start)
    log.Printf("%s took %s", name, elapsed)
}


func getFileInput(inputFile string) (string, error) {
	rawData, err := os.ReadFile(inputFile)
	if err != nil {
		return "", err
	}

	return string(rawData), nil
}

func multiplyMatchNums(matchGroup []string) (int, error) {
	a, errA := strconv.Atoi(matchGroup[1])
	b, errB := strconv.Atoi(matchGroup[2])

	if errA != nil {
		return 0, errA
	}
	if errB != nil {
		return 0, errB
	}

	return a * b, nil
}


func partOne(dataString string) (result int, err error) {
	defer timeTrack(time.Now(), "part one")

	re := regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)`)
	matches := re.FindAllStringSubmatch(dataString, -1)
	for _, matchGroup := range matches {
		mult, err := multiplyMatchNums(matchGroup)
		if err != nil {
			return 0, err
		}

		result += mult
	}
	return
}


func partTwo(dataString string) (result int, err error) {
	defer timeTrack(time.Now(), "part one")

	re := regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)`)
	matches := re.FindAllStringSubmatch(dataString, -1)
	enabled := true
	for _, matchGroup := range matches {
		switch matchGroup[0] {
		case "do()":
			enabled = true
		case "don't()":
			enabled = false
		default:
			if enabled {
				mult, err := multiplyMatchNums(matchGroup)
				if err != nil {
					return 0, err
				}

				result += mult
			}
		}
	}

	return
}


func main() {
	flag.Parse()
	
	var inputFile string
	if *isTest {
		inputFile = "input_test.txt"
		fmt.Println("Running 2024/3 solution on test input.")
	} else {
		inputFile = "input.txt"
		fmt.Println("Running 2024/3 solution on full input.")
	}

	data, err := getFileInput(inputFile)
	if err != nil {
		fmt.Printf("Error reading file:\n> %v", err)
		return
	}

	partOneSolution, err := partOne(data)
	if err != nil {
		fmt.Printf("Error solving part one:\n> %v", err)
		return
	}
	fmt.Printf("Part 1: %v\n", partOneSolution)

	partTwoSolution, err := partTwo(data)
	if err != nil {
		fmt.Printf("Error solving part two:\n> %v", err)
		return
	}

	fmt.Printf("Part 2: %v\n", partTwoSolution)
}

File: 2024/3/solution.py
"""https://adventofcode.com/2024/day/3"""
import argparse
import re
from typing import Callable
import timeit


def timed(f: Callable) -> Callable:
    """Decorator to time the execution of a function."""
    def wrapper(*args, **kwargs):
        start_time = timeit.default_timer()
        result = f(*args, **kwargs)
        end_time = timeit.default_timer()
        print(f"Function {f.__name__} took {end_time - start_time:.6f} seconds")
        return result
    return wrapper


@timed
def part_one(data: str) -> int:
    r = re.compile(r"mul\((\d{1,3}),(\d{1,3})\)")

    return sum(int(m.group(1)) * int(m.group(2)) for m in r.finditer(data))


@timed
def part_two(data: list[str]):
    r = re.compile(r"mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)")
    res = 0
    enabled = True

    for m in r.finditer(data):
        match m.group(0):
            case "do()":
                enabled = True
            case "don't()":
                enabled = False
            case _ if enabled:
                res += int(m.group(1)) * int(m.group(2))

    return res


parser = argparse.ArgumentParser(description="Solution for Advent of Code 3/2024.")
parser.add_argument("-t", "--test", action="store_true", help="use test input")

if __name__ == "__main__":
    args = parser.parse_args()
    filename: str
    if args.test:
        filename = "./input_test.txt"
        print("Running 2024/3 solution on test input.")
    else:
        filename = "./input.txt"
        print("Running 2024/3 solution on full input.")

    with open(filename) as f:
        raw_data = f.read()

    first_answer = part_one(raw_data)
    print(f"PART ONE: The answer to part one is equal to {first_answer}.")
    second_answer = part_two(raw_data)
    print(f"PART TWO: The answer to part two is equal to {second_answer}.")
